Doc 11827 mobile vector search (#848)
couchbaselabs/docs-couchbase-lite: modules/android/pages/vector-search-api-reference.adoc

= Vector Search API Reference
:page-aliases: 
ifdef::show_edition[:page-edition: {release}]
ifdef::prerelease[:page-status: {prerelease}]
:page-role:

> :description: API reference for Vector Search in Couchbase Lite.

Reviewer: Don't forget to expand this and include it in the body of the page with: 

[abstract]
{description}

:keywords: edge AI api swift ios macos apple vector search generative

== Vector Index API

> You can use the Vector Index API's core functionality to efficiently perform similarity search at scale on Vector Indexes.

Reviewer: ```suggestion
You can use the Vector Index API's core functionality to efficiently perform similarity search at scale on Vector Search indexes.
```

Author: Done!

> You can make use of new qurry functions such as `vector_match()` to query indexed vectors.

Reviewer: ```suggestion
You can make use of new query functions such as `vector_match()` to query indexed vectors.
```

Author: Done!

=== Vector Encoding

Vector encoding is the process of representing data points such as images, words, or documents as vectors in high-dimensional space where each dimension represents a different attribute of the data. 

> Encoding is a crucial step in preparing data for similarity or nearest neighbour search operations.

Reviewer: ```suggestion
Encoding is a crucial step in preparing data for similarity or nearest-neighbor search operations.
```

Watch the British spelling.

Reviewer 2: Not sure if `Vector Encoding` mentioned in this section is about the vector encoding methods used in the vector index or not. If it is, Vector Encoding is the method or algorithm that compresses the vectors stored in the vector index. The encoding makes the index size smaller and reduce the I/P during index and query time, however, on the flip side, encoding could make the distance computation and query result less inaccurate.

We support 3 types of encoding in our API: None (No encoding, No data-loss ), SQ and PQ. Here are a bit of info about the SQ and PQ encoding:

https://github.com/couchbaselabs/mobile-vector-search/blob/main/docs/Tuning.md#2-vector-encoding

> Couchbase Lite support three encoding types:

Reviewer 3: There appears to be some confusion around vector encoding. Please catchup with Iveta on this
What you describe here is how Couchbase Lite vector indexes are encoded. 

At a high level, this is what happens
- Data is translated to vector of floats
- Vector index is created on the vectors. You use encoding to save space when storing indexes.

> Couchbase Lite support three encoding types:

Reviewer: ```suggestion
Couchbase Lite supports three encoding types:
```

}

----

IMPORTANT: The beta release currently only supports the format of vectors as an array of numbers.

=== Default Value Constants

[source, c]
----

public class VectorIndexConfiguration {
  /* Scalar Quantizer */
  public static let defaultEncoding = VectorEncoding.scalarQuantizer()

  /* Squared Euclidean Distance */
  public static let defaultMetric = VectorMetric.euclidean   
}

----

=== Extension Path Configuration

NOTE: The extension configuration is only necessary for the CBL-iOS MacOS and CBL-C platforms as dynamic extension libraries may be at arbitrary locations.

[source, c]
----

public class CouchbaseLite {
  /* Registers a directory path to load extension libraries from. Must 
     be called before using CouchbaseLite but only when using extensions */
  public static func setExtensionPath(extensionPath: String)
}

----

== Vector SQL++ Functions

Couchbase Lite currently supports two SQL++ functions, `vector_match()` and `vector_distance()`.

=== Vector Match

==== `vector_match(vectorIndexIdentifier, targetVectorExpr, [limit = 3])`

[cols = "3*"]
|===
|Parameter |Is Required |Description

|vectorIndexIdentifier
|Required

> |The name of the vector index to perform the vector search on.

Reviewer: ```suggestion
|The name of the vector index to search.
```

|targetVectorExpr
|Required
|The target vector expression that returns a vector in the form of an array of numbers.
|limit
|Optional
|The limit number of the returned matched results.
The maximum number allowed is 10000. An error will be returned when creating a query with a limit greater than 10000.

|===

> This function performs vector search against a specific vector index identifier for the specified vector expression.

Reviewer: ```suggestion
This function performs a vector search against a specific vector index identifier for the specified vector expression.
```
Do we want to be capitalizing Vector Search at all in this description? Try to keep consistent.

If the specified index does not exist, an error will occur on creation of the query.

> The matched vectors will be returned up to the specified limit number, if the limit is not specified then the default value will be used. 

Reviewer: ```suggestion
The matched vectors will be returned up to the specified limit number. 
If you do not specify a limit, then Couchbase Lite uses a default value of `3`. 
```

The returned vectors are sorted by their distance values in ascending order by default.

> NOTE: The default value for the `limit` parameter is 3.

Reviewer: ```suggestion

```

[source, sqlpp]
----
SELECT name 
FROM photos 
WHERE vector_match(photos-index, [0.1, 0.5, 1.2])

----

[source, sqlpp]
----

SELECT name 
FROM photos
WHERE vector_match(photos-index, [0.1, 0.5, 1.2]) AND city = "San Francisco"

----

=== Vector Distance

> 

Reviewer: Are you going to add any preamble here?

Reviewer: You should try to avoid stacking headings with no content in between.

==== `vector_distance(vectorIndexIdentifier)`

[cols = "3*"]
|===
|Parameter |Is Required |Description

|vectorIndexIdentifier
|Required
|The name of the vector index.

|===

> This function returns the distance between the target vector specified in the `vector_match()` function and the matched vector in the specified vector index based on the distance metric set in the index configuration .

Reviewer: ```suggestion
This function returns the distance between the target vector specified in the `vector_match()` function and the matched vector in the specified vector index, based on the distance metric set in the index configuration.
```

[source, sqlpp]
----

SELECT name, vector_distance(photos-index) 
FROM photos
WHERE vector_match(photos-index, [0.1, 0.5, 1.2])

----

[source, sqlpp]
----

SELECT name WHERE vector_match(photos-index, [0.1, 0.5, 1.2]) 
FROM photos
ORDER BY vector_distance(photos-index) DSC

----

IMPORTANT: Similar to the Full Text Search `match()` function, `vector_match()` can only be called alone or at the top level `AND` expression.

=== Generate Vectors with Prediction

You can use two methods to generate vectors in Couchbase Lite:

> . You can call a Machine Learning(ML) model, and embed the generated vectors inside the documents.

Reviewer: ```suggestion
. You can call a Machine Learning (ML) model, and embed the generated vectors inside the documents.
```

> . You can use the `prediction()` function to generate vectors to be indexed for each document at the indexing time.

Reviewer: ```suggestion
. You can use the `prediction()` function to generate vectors to be indexed for each document at indexing time.
```

Below is an example configuration.

[source, c]
----

let config = VectorIndexConfiguration(expression: "prediction(photo-model, photo).vector", dimensions: 256, centroids: 30)
try collection.createIndex(withName: "photo-index", config: config)

----

You can use less storage by using the `prediction()` function as the encoded vectors will only be stored in the index. 
However, the index time will be longer as indexing is occurring at run time.

[IMPORTANT]
--

> In the beta release. Updating the index for the documents whose returned vectors from the `prediction()` were missing on calling the ML model is not possible unless you perform the following actions:

Reviewer: Unclear on what you're trying to say here?

* Update the documents

* Regenerate the index

--

== See Also

* xref:swift:vector-search.adoc[Vector Search]

* xref:swift:working-with-vector-search.adoc[Working with Vector Search]

* xref:swift:fts.adoc[Full Text Search]

> 

Reviewer 3: General- we must not duplicate what is in API reference guides here (expectation is that we will publish those when we go Beta). We must only include code snippets describing usage and then link to API reference guide for correct API definitions.

= Vector Search API Reference
:page-status: Beta
:page-aliases: 
ifdef::show_edition[:page-edition: {release}]
ifdef::prerelease[:page-status: {prerelease}]
:page-role:
:description: API reference for Vector Search in Couchbase Lite.
:keywords: edge AI api swift ios macos apple vector search generative

== Vector Index API

You can use the Vector Index API's core functionality to efficiently perform similarity search at scale on Vector Indexes.
You can make use of new qurry functions such as `vector_match()` to query indexed vectors.

=== Vector Encoding

Vector encoding is the process of representing data points such as images, words, or documents as vectors in high-dimensional space where each dimension represents a different attribute of the data. 
Encoding is a crucial step in preparing data for similarity or nearest neighbour search operations.

Couchbase Lite support three encoding types:

* No encoding - 4 bytes per dimension, large size with no data loss

> * Scalar Quantizer - The default setting.

Reviewer 2: The default encoding is Scalar Quantizer (8 bits).

Author: Done, thank you!
